src/hooks/useFetch.js

```diff
@@ -0,0 +1,72 @@
+import { useEffect, useState } from 'react';
+
+const fetchUrl = (url, after, count, limit = 100, time = 'year') => {
+  const urlParams = `t=${time}&limit=${limit}${
+    count && after ? `&after=${after}&count${count}` : ''
+  }`;
+  return `${url}/top.json?${urlParams}`;
+};
+
+const makeRequest = async (url) => {
+  const res = await fetch(url);
+
+  if (!res.ok) {
+    throw new Error('Unable to fetch data');
+  }
+
+  const data = await res.json();
+  return data;
+};
+
+const fetchPosts = async (
+  url,
+  posts = [],
+  count = null,
+  after = null,
+  numRequests = 500,
+) => {
+  if (count >= numRequests) {
+    return posts;
+  }
+
+  const currentUrl = fetchUrl(url, after, count);
+
+  const data = await makeRequest(currentUrl);
+
+  const recordsRetrieved = data.data.dist;
+  posts.push(...data.data.children);
```

Comment:
I'm surprised that ESLint is not complaining. Here you're mutating the posts parameter. This can be dangerous since the posts will be changed outside the scope of this function as well. Here is an example that might be confusing

```
const posts = [];
const data = await fetchPosts(url, posts);
console.log(posts);
```

Now the posts variable would not be empty anymore but contain the complete data set.

To avoid this kind of side effect of a function you can simply create a new variable

```
const newPosts = posts.concat(data.data.children);
```